model/modeldecoder/v2/simdjson.go

```
@@ -0,0 +1,593 @@
+	iter.Advance()
+	typ, _, err := iter.Root(tmpIter)
+	if err != nil {
+		return err
+	}
+	if typ != simdjson.TypeObject {
+		err = errors.New("element not an object")
+		return err
+	}
+	obj, err = tmpIter.Object(obj)
+	if obj == nil {
+		return errors.New("object not found")
+	}
+	return err
+}
+
+func keyObject(obj *simdjson.Object, key string) error {
+	elem := obj.FindKey(key, nil)
+	if elem == nil || elem.Type != simdjson.TypeObject {
+		return fmt.Errorf("object with key %v not found", key)
+	}
+	obj, err := elem.Iter.Object(obj)
+	if err != nil {
+		return err
+	}
+	if obj == nil {
+		return errors.New("object not found")
+	}
+	return err
+}
+
+func decodeElements(obj *simdjson.Object, key string, out interface{}) error {
+	elems := elemsPool.Get().(*simdjson.Elements)
+	defer elemsPool.Put(elems)
+	_, err := obj.Parse(elems)
+	if err != nil {
+		return err
+	}
+	for _, elem := range elems.Elements {
+		s := elem.Name
+		if key != "" {
+			s = fmt.Sprintf("%s.%s", key, s)
+		}
+		switch elem.Type {
+		case simdjson.TypeObject:
+			//TODO(simitt): improve this
+			if elem.Name == "labels" {
+				if err := elementToDecoderModel(s, elem, out); err != nil {
+					return err
+				}
+			} else {
+				elemObj, err := elem.Iter.Object(nil)
+				if err != nil {
+					return err
+				}
+				if err := decodeElements(elemObj, s, out); err != nil {
+					return err
+				}
+			}
+		default:
+			if err := elementToDecoderModel(s, elem, out); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+//TODO(simitt): to be generated from json tags and datatypes
+func elementToDecoderModel(key string, elem simdjson.Element, out interface{}) error {
+	switch typ := out.(type) {
+	case *Metadata:
+		return elementToMetadata(key, elem, typ)
+	}
+	return nil
+}
+
+func elementToMetadata(key string, elem simdjson.Element, m *Metadata) error {
+	switch key {
+	case "cloud.account.id":
+		if elem.Type == simdjson.TypeString {
```

If we silently ignore other types, how will we enforce type validation? Perhaps we should check `elem.Type != simdjson.TypeNull`, and then just call `elem.Iter.String()` which will return an error if it's not a string?

---

good point, will change that